### DISCUSSION TITLE
Switch Matrix modification for custom tiles

### DISCUSSION THREAD

**Discussion by dishant1602** [FPGA-Research/FABulous#259] [Q&A]
I want to create a custom tile(which has a BEL of AND_gate.v file), but I am facing issue when modifying this LUT4AB switch matrix. Looks like there are alot of connections and I can't figure out which lines exactly to modify the switch_matrix.list file in order to simply modify it for the purpose of two input AND gate.

**Comment by EverythingElseWasAlreadyTaken** [comment]
If you are just taking the LUT4AB switch matrix, you can simply use all connections beginning with an L, since they are the representation of the LUT4c_frame_config_dffesr BEL ports in the switch matrix. 

``` 
fabric.csv
...
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LA_,,,,,,,,,,,,,,,,
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LB_,,,,,,,,,,,,,,,,
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LC_,,,,,,,,,,,,,,,,
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LD_,,,,,,,,,,,,,,,,
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LE_,,,,,,,,,,,,,,,,
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LF_,,,,,,,,,,,,,,,,
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LG_,,,,,,,,,,,,,,,,
BEL,./Tile/LUT4AB/LUT4c_frame_config_dffesr.v,LH_,,,,,,,,,,,,,,,,
...
```

```
LUT4c_frame_config_dffesr.v Ports
... 
input I0; // LUT inputs
input I1;
input I2;
input I3;
output O; // LUT output (combinatorial or FF)
input Ci; // carry chain input
output Co; // carry chain output
input SR; // SHARED_RESET
input EN; // SHARED_ENABLE
...
```

All BEL ports are added with their defined prefix to the switch matrix. 
This means for BEL LA_ you'll find the ports LA_I0, LA_I1 ... LA_EN in the switch matrix. 
Check the docs for the list file syntax: https://fabulous.readthedocs.io/en/latest/fabric_definition.html#switch-matrix

So if you are replacing one or more LUT4c_frame_config_dffesr BEL with some other BEL, you have to make sure, that you replace all occurrences of the old BEL ports with your new BEL ports or replace them with some other Ports or kick out the whole multiplexer. Also, you have to look out for the carry chain, reset and enable ports, since are implemented a bit differently in the switch matrix. Check the comments in the list file for this.

**Reply by Sacoder0000** [reply]
Thank you!
Changing SwitchMatrix.list is a very complicated and tedious task for users since we don't understand how the original switchmatrix works. I think it will be very convenient if there are more detailed explanatory documents, such as SB, CB, routing segement.

**Reply by KelvinChung2000** [reply]
We decided not to use abstraction like SB and CB as VPR does because we would like fine-grain control on switch matrix construction. As a result, we are describing each of the Muxes in the switch matrix using the list file. An analogy would be FABulous, which is an FPGA assembler, and VPR, which is more like a compiler. 

However, I agree that being an assembler does not mean that we cannot have a hierarchy to aid implementation. The list file system needs a rework to make it more intuitive while maintaining the ability to fine-grain control all the Muxes.

**Reply by Sacoder0000** [reply]
Thank you for all your efforts! This is truly a remarkable project. I look forward to seeing further improvements!
